### PR TITLE
examples for int8

### PR DESCRIPTION
**Pull request by i8run**
## What changes were proposed in this pull request?

This is an example of how to use mkldnn int8. There're two steps, use Quantize to generate the scales first and save the new model. And than you can use the quantized model as usual.

## How was this patch tested?

Manual test on VGG-16 and ResNet-50.

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      val model = Module.loadModule[Float](param.model).toGraph()
+
+      model.evaluate()
+
+      model.setInputDimMask(0)
+      model.setOutputDimMask(0)
+      model.setWeightDimMask(1)
+
+      samples.foreach { sample =>
+        model.forward(sample)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala]: why need forward here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import com.intel.analytics.bigdl.dataset.image.CropCenter
+import com.intel.analytics.bigdl.dataset.{DataSet, Sample, SampleToMiniBatch}
+import com.intel.analytics.bigdl.nn.Module
+import com.intel.analytics.bigdl.transform.vision.image.augmentation.{ChannelScaledNormalizer, RandomCropper, RandomResize}
+import com.intel.analytics.bigdl.transform.vision.image.{ImageFeature, ImageFrameToSample, MatToTensor, PixelBytesToMat}
+import com.intel.analytics.bigdl.utils.Engine
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SparkContext
+
+object Quantize {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala]: I think this example is confused for users. Not clear about how to get scales.

Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala]: Yes. This only generate the scales.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala (hunk 1 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/int8/Quantize.scala]: It will generate output. Legacy code for dnn model. I'll delete it.

**Comment by wzhongyuan** [comment]
/example/mkldnn/int8 seems better ? we may have other examples for mkldnn in future

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import com.intel.analytics.bigdl.dataset.image.CropCenter
+import com.intel.analytics.bigdl.dataset.{DataSet, Sample, SampleToMiniBatch}
+import com.intel.analytics.bigdl.nn.Module
+import com.intel.analytics.bigdl.transform.vision.image.augmentation.{ChannelScaledNormalizer, RandomCropper, RandomResize}
+import com.intel.analytics.bigdl.transform.vision.image.{ImageFeature, ImageFrameToSample, MatToTensor, PixelBytesToMat}
+import com.intel.analytics.bigdl.utils.Engine
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SparkContext
+
+object GenerateInt8Scales {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala]: add comments

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        .set("spark.akka.frameSize", 64.toString)
+        .set("spark.task.maxFailures", "1")
+      val sc = new SparkContext(conf)
+      Engine.init
+
+      // We set partition number to be node*core, actually you can also assign other partitionNum
+      val partitionNum = Engine.nodeNumber()
+      val imageFrame = DataSet.SeqFileFolder.filesToImageFrame(param.folder, sc, 1000,
+        partitionNum = Option(partitionNum))
+      val transformer =
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala]: add comments that the transformer should be the same as that in validation during training

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      }
+
+      val toMiniBatch = SampleToMiniBatch[Float](param.batchSize)
+      val samples = evaluateDataSet
+        .repartition(1)
+        .mapPartitions(toMiniBatch(_))
+        .take(1)
+        .map(_.getInput().toTensor[Float])
+
+      val model = Module.loadModule[Float](param.model).toGraph()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala]: why explicitly call `toGraph`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      model.setWeightDimMask(1)
+
+      samples.foreach { sample =>
+        model.forward(sample)
+        model.calcScales(sample)
+      }
+
+      model.clearState()
+
+      model.saveModule(path = param.model.concat(".quantized"), overWrite = true)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala]: By convention, all pre-tained model file name should have suffix as `.bigdl`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# How to Use Lower Numerical Precision Deep Learning Inference in BigDL
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/README.md]: => Inference acceleration with MKL-DNN  Low Numerical Precision (Int8) computation ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+  def main(args: Array[String]): Unit = {
+    testParser.parse(args, new TestParams()).foreach { param =>
+      val conf = Engine.createSparkConf().setAppName("Quantize the model")
+        .set("spark.akka.frameSize", 64.toString)
+        .set("spark.task.maxFailures", "1")
+      val sc = new SparkContext(conf)
+      Engine.init
+
+      // We set partition number to be node*core, actually you can also assign other partitionNum
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala]: is it true ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        .repartition(1)
+        .mapPartitions(toMiniBatch(_))
+        .take(1)
+        .map(_.getInput().toTensor[Float])
+
+      val model = Module.loadModule[Float](param.model).toGraph()
+
+      model.evaluate()
+
+      model.setInputDimMask(0)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/GenerateInt8Scales.scala]: why need to specify it?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/TestImageNet.scala [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import com.intel.analytics.bigdl.optim._
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric._
+import com.intel.analytics.bigdl.utils._
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SparkContext
+
+/**
+ * This example is to evaluate trained resnet50 with imagenet data and get top1 and top5 accuracy
+ */
+object TestImageNet {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/TestImageNet.scala]: ImageNetInference ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/TestImageNet.scala [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import com.intel.analytics.bigdl.models.resnet.ImageNetDataSet
+import com.intel.analytics.bigdl.nn.Module
+import com.intel.analytics.bigdl.optim._
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric._
+import com.intel.analytics.bigdl.utils._
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SparkContext
+
+/**
+ * This example is to evaluate trained resnet50 with imagenet data and get top1 and top5 accuracy
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/TestImageNet.scala]: This example demonstrates how to evaluate pre-trained resnet-50 with ImageNet dataset using Int8

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/ImageNetInference.scala [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+      val name = prefix.concat(".quantized").concat(suffix)
+      logger.info(s"Save the quantized model $name ...")
+      // it will force overWrite the existed model file
+      model.saveModule(name, overWrite = true)
+      logger.info(s"Save the quantized model $name done.")
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
+    testParser.parse(args, TestParams()).foreach(param => {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/ImageNetInference.scala]: why not put the Util into the same class ? no much logic for that class

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/README.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
+# Inference acceleration with MKL-DNN Low Numerical Precision (Int8) Computating
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/README.md]: Sorry, typo computating -> computing

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/ImageNetInference.scala (hunk 12 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnnint8/ImageNetInference.scala]: I'll put it into a single file.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+      val imageFrame = DataSet.SeqFileFolder.filesToImageFrame(param.folder, sc, 1000,
+        partitionNum = Option(partitionNum))
+
+      // the transformer is the same as as that in validation during training
+      val evaluationSet = ImageNetDataSet.valDataSet(param.folder,
+        sc, 224, param.batchSize).toDistributed().data(train = false)
+
+      val samples = evaluationSet
+        .repartition(1) // repartition (shuffle) will have better accuracy
+        .take(param.numOfBatch) // only split one or some batches to sample
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala]: check data correction

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        .map(_.getInput().toTensor[Float])
+
+      // Currently, we only support the graph model, so we add a `toGraph`
+      // if the model is already graph, you can need not to it.
+      val model = Module.loadModule[Float](param.model).toGraph()
+
+      model.setInputDimMask(0)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala]: extract scaled operation to a method, and user can call this method directly

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala (hunk 14 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala]: the output will be same when numOfBatch more than 1 and we can't only split some of the samples too. I only just change it back to 1. user can use the batch size to control them.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala (hunk 15 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/mkldnn/int8/GenerateInt8Scales.scala]: Done.

**Comment by i8run** [comment]
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/484/. Jenkins passed.

**Pull request merged**